File: backend/app/schemas/validation.py
"""
Pydantic DTO Schemas for Empirical Validation and Benchmarking
"""

from datetime import datetime
from typing import Any, Dict, List, Optional
import uuid
from pydantic import BaseModel, ConfigDict, Field


class ValidationRunCreateRequest(BaseModel):
    name: Optional[str] = None
    tags: Optional[List[str]] = None
    max_cases: Optional[int] = None
    notes: Optional[str] = None


class ValidationCaseResultResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    validation_run_id: uuid.UUID
    test_case_id: str
    title: str
    category: str
    document_type: str
    quality_level: str
    expected_result_json: Dict[str, Any]
    actual_result_json: Dict[str, Any]
    is_correct: bool
    error_type: str
    error_reason: Optional[str] = None
    ocr_correct: bool
    ocr_accuracy: float
    classification_correct: bool
    extraction_correct: bool
    compliance_correct: bool
    rag_correct: bool
    processing_time_ms: float
    manual_baseline_sec: float
    details_json: Optional[Dict[str, Any]] = None
    created_at: datetime


class ValidationRunResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    organization_id: Optional[uuid.UUID] = None
    name: str
    dataset_version: str
    engine_versions: Optional[Dict[str, Any]] = None
    status: str
    total_cases: int
    passed_cases: int
    failed_cases: int
    ocr_accuracy: float
    classification_accuracy: float
    field_extraction_accuracy: float
    compliance_accuracy: float
    true_positives: int
    true_negatives: int
    false_positives: int
    false_negatives: int
    precision: float
    recall: float
    f1_score: float
    false_positive_rate: float
    false_negative_rate: float
    rag_retrieval_accuracy: float
    rag_citation_accuracy: float
    average_processing_time_ms: float
    average_manual_time_sec: float
    time_reduction_percentage: float
    summary_json: Optional[Dict[str, Any]] = None
    notes: Optional[str] = None
    started_at: Optional[datetime] = None
    completed_at: Optional[datetime] = None
    created_at: datetime
    updated_at: datetime


class ValidationRunListResponse(BaseModel):
    items: List[ValidationRunResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class ValidationCaseListResponse(BaseModel):
    items: List[ValidationCaseResultResponse]
    total: int
    page: int
    page_size: int
    total_pages: int


class ValidationDatasetCaseResponse(BaseModel):
    id: str
    title: str
    category: str
    document_type: str
    quality_level: str
    expected_doc_type: str
    expected_compliance_status: str
    manual_baseline_sec: float


class ValidationPPTSummaryResponse(BaseModel):
    slide_title: str
    dataset_overview: Dict[str, Any]
    performance_metrics: Dict[str, str]
    speed_and_efficiency_gains: Dict[str, str]
    key_takeaways: List[str]
    observed_limitations: List[str]

File: frontend/src/types/validation.ts
/**
 * Empirical Validation and Benchmarking Types
 */

export interface ValidationRun {
  id: string;
  organization_id?: string | null;
  name: string;
  dataset_version: string;
  engine_versions?: Record<string, any> | null;
  status: "PENDING" | "RUNNING" | "COMPLETED" | "FAILED";
  total_cases: number;
  passed_cases: number;
  failed_cases: number;

  // Accuracies
  ocr_accuracy: number;
  classification_accuracy: number;
  field_extraction_accuracy: number;
  compliance_accuracy: number;

  // Confusion Matrix
  true_positives: number;
  true_negatives: number;
  false_positives: number;
  false_negatives: number;

  // Statistical Rates
  precision: number;
  recall: number;
  f1_score: number;
  false_positive_rate: number;
  false_negative_rate: number;

  // RAG
  rag_retrieval_accuracy: number;
  rag_citation_accuracy: number;

  // Timing
  average_processing_time_ms: number;
  average_manual_time_sec: number;
  time_reduction_percentage: number;

  summary_json?: {
    confusion_matrix?: {
      true_positives: number;
      true_negatives: number;
      false_positives: number;
      false_negatives: number;
    };
    quality_correlation?: Record<string, { count: number; avg_ocr_accuracy: number }>;
    category_breakdown?: Record<string, { total: number; accuracy: number; avg_ocr: number; avg_extraction: number }>;
    document_type_breakdown?: Record<string, { total: number; accuracy: number }>;
    load_performance?: Record<string, { bids_count: number; estimated_total_time_sec: number; success_rate: number }>;
  } | null;

  notes?: string | null;
  started_at?: string | null;
  completed_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface ValidationCaseResult {
  id: string;
  validation_run_id: string;
  test_case_id: string;
  title: string;
  category: string;
  document_type: string;
  quality_level: string;
  expected_result_json: Record<string, any>;
  actual_result_json: Record<string, any>;
  is_correct: boolean;
  error_type: string;
  error_reason?: string | null;
  ocr_correct: boolean;
  ocr_accuracy: number;
  classification_correct: boolean;
  extraction_correct: boolean;
  compliance_correct: boolean;
  rag_correct: boolean;
  processing_time_ms: number;
  manual_baseline_sec: number;
  details_json?: Record<string, any> | null;
  created_at: string;
}

export interface ValidationRunListResponse {
  items: ValidationRun[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface ValidationCaseListResponse {
  items: ValidationCaseResult[];
  total: number;
  page: number;
  page_size: number;
  total_pages: number;
}

export interface ValidationPPTSummary {
  slide_title: string;
  dataset_overview: Record<string, any>;
  performance_metrics: Record<string, string>;
  speed_and_efficiency_gains: Record<string, string>;
  key_takeaways: string[];
  observed_limitations: string[];
}
